mithril-common/src/cardano_block_scanner/chain_reader_block_streamer.rs

```diff
@@ -0,0 +1,320 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use slog::{debug, Logger};
+use tokio::sync::Mutex;
+
+use crate::cardano_block_scanner::BlockStreamer;
+use crate::cardano_block_scanner::ChainScannedBlocks;
+use crate::chain_reader::{ChainBlockNextAction, ChainBlockReader};
+use crate::entities::BlockNumber;
+use crate::entities::ChainPoint;
+use crate::StdResult;
+
+/// The action that indicates what to do next with the streamer
+#[derive(Debug, Clone, PartialEq)]
+enum BlockStreamerNextAction {
+    /// Use a [ChainBlockNextAction]
+    ChainBlockNextAction(ChainBlockNextAction),
+    /// Skip to the next action
+    SkipToNextAction,
+}
+
+/// The maximum number of roll forwards during a poll
+const MAX_ROLL_FORWARDS_PER_POLL: usize = 100;
+
+/// [Block streamer][BlockStreamer] that streams blocks with a [Chain block reader][ChainBlockReader]
+pub struct ChainReaderBlockStreamer {
+    chain_reader: Arc<Mutex<dyn ChainBlockReader>>,
+    from: ChainPoint,
+    until: BlockNumber,
+    max_roll_forwards_per_poll: usize,
+    logger: Logger,
+}
+
+#[async_trait]
+impl BlockStreamer for ChainReaderBlockStreamer {
+    async fn poll_next(&mut self) -> StdResult<Option<ChainScannedBlocks>> {
+        debug!(self.logger, "ChainReaderBlockStreamer polls next");
+
+        let chain_scanned_blocks: ChainScannedBlocks;
+        let mut roll_forwards = vec![];
+        loop {
+            let block_streamer_next_action = self.get_next_chain_block_action().await?;
+            match block_streamer_next_action {
+                Some(BlockStreamerNextAction::ChainBlockNextAction(
+                    ChainBlockNextAction::RollForward { parsed_block },
+                )) => {
+                    roll_forwards.push(parsed_block);
+                    if roll_forwards.len() >= self.max_roll_forwards_per_poll {
+                        return Ok(Some(ChainScannedBlocks::RollForwards(roll_forwards)));
+                    }
+                }
+                Some(BlockStreamerNextAction::ChainBlockNextAction(
+                    ChainBlockNextAction::RollBackward { rollback_point },
```

Comment:
Maybe we could add a log in case of a rollback? Since it's a rare case, we would want to track it?

Reply:
This is alreday the case in the `get_next_chain_block_action` function: 
https://github.com/input-output-hk/mithril/blob/03aafd9b8494ee9f656f9c716660b31e4cbc6eed/mithril-common/src/cardano_block_scanner/chain_reader_block_streamer.rs#L125